### PR TITLE
Football v1

### PR DESCRIPTION
**Pull request by cyfra**
Adding Google Research Football to the list of Kaggle Environments.

### CODE REVIEW

File: kaggle_environments/envs/football/test_football.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    assert json["specification"]["reward"]["type"] == ["number", "null"]
+
+def clear_minimap(state):
+  state = copy.deepcopy(state)
+  for entry in state:
+    entry.observation.minimap = []
+  return state
+
+
+def test_single_agent():
```
Comment (harrisse) [inline kaggle_environments/envs/football/test_football.py:30]: Solid start to testing. I'm guessing there are a number of edge cases about how units collide, out of bounds, etc that are worth testing as well. Is there another test suite for research football that already covers these cases?

File: kaggle_environments/envs/football/football.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+# Global dictionary with active environments.
+m_envs = {}
+
+def interpreter(state, env):
+  global m_envs
+  if (env.id not in m_envs) or env.done:
+    print("Staring a new environment %s: with scenario: %s" % (env.id, env.configuration.scenario_name))
+    m_envs[env.id] = football_env().create_environment(env_name=env.configuration.scenario_name,
+                                            stacked=False,
+                                            logdir='/tmp/football',
```
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/football/football.py:64]: Also same caveat to the internal info field in the specification.

File: kaggle_environments/envs/football/football.ipynb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        "colab": {}
+      },
+      "source": [
+        "# Installing Google Football environment.\n",
+        "\n",
+        "!apt-get update\n",
+        "!apt-get install libsdl2-gfx-dev libsdl2-ttf-dev\n",
+        "\n",
+        "# Make sure that the Branch in git clone and in wget call matches !!\n",
+        "!git clone -b v2.0.6 https://github.com/google-research/football.git\n",
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.ipynb:50]: yes, we'll need to add those deps there.
let's chat offline about the possible options.

File: kaggle_environments/envs/football/football.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  drawLine({x1:rightX, y1:topY, x2:rightX, y2:bottomY, style:{lineWidth:unit, strokeStyle:"#ffffff"}});
+
+  // drawCell({x:10, y:10, color:"#0000ff"});
+
+  const minimap = environment.steps[step][0].observation.minimap;
+  var x,y,z;
+  var i = 0;
+  var colors = ["black", "red", "yellow", "purple"];
+
+  for (y = 0; y<72; y++) {
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.js:89]: I've refactored this code (as I've also changed the representation).

File: kaggle_environments/envs/football/football.json [hunk 5]
```diff
@@ -0,0 +1,61 @@
+{
+    "name": "football",
+    "agents": [2],
+    "configuration": {
+      "episodeSteps": 30,
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.json:5]: This is just an example for now.

But we should also discuss this part - as in this environment, the number of steps is configured within the 'scenario'.

File: kaggle_environments/envs/football/football.json [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      "team_2": {
+          "description": "Number of players that other agent controls. If set to 0 - the second agent will always have to return 0 as action.",
+          "type": "integer",
+          "minimum": 0,
+          "maximum": 10,
+          "default": 0
+      }
+    },
+    "reward": {
+      "description": "1.0 for scored goal, -1.0 for lost goal (it is given only when the goal is scored, and changes back to 0 afterwards)",
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.json:27]: It can depend on us.
The underlying environment is giving a +1, -1 reward only after the goal was scored (and then rewards go back to 0) - this makes training more stable.

File: kaggle_environments/envs/football/football.json [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    "info": {
+      "debug_info": {
+          "description": "Human readable information passed from the system.",
+          "type": "string"
+      }
+    },
+
+    "observation": {
+        "minimap": {
+          "description": "72x96x4 tensor. Channels: your team, opponent, ball, active_player ",
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.json:40]: I've changed the format to send 'raw' observations, that people will be able to convert to minimap if needed.

(but we should still discuss on how much info we should put inside this schema vs inside the README file).

File: kaggle_environments/envs/football/football.json [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      }
+    },
+    "reward": {
+      "description": "1.0 for scored goal, -1.0 for lost goal (it is given only when the goal is scored, and changes back to 0 afterwards)",
+      "type": "number"
+    },
+
+    "info": {
+      "debug_info": {
+          "description": "Human readable information passed from the system.",
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.json:33]: Currently, I print things like "you won because your enemy did a wrong move" etc.

File: kaggle_environments/envs/football/football.json [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        },
+        "controlled_players": {
+            "description": "number of players that agent controls",
+            "type": "number",
+            "minimum": 0,
+            "maximum": 10
+        }
+    },
+    "action": {
+      "description": "A Button to press for each player that agent controls: 0-idle, 1-left, 2-top_left, etc. ",
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.json:53]: ack - I'll prepare a separate page with documentation for actions.

File: kaggle_environments/envs/football/football.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
+import importlib
+import json
+from os import path
+import numpy as np
+
+def renderer(state, env):
+  return "inline rendering not supported."
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.py:7]: I guess we could put here the final score? (or maybe also minutes where goals were scored etc?)

File: kaggle_environments/envs/football/football.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        state[agent].reward = rew[agent].item()
+
+
+def update_state_on_invalid_action(bad_agent, good_agent, message):
+  bad_agent.status = "INVALID"
+  bad_agent.reward = -100
+  bad_agent.info.debug_info = message
+
+  good_agent.status = "DONE"
+  good_agent.reward = 100
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.py:48]: nope - just 'large value'.

File: kaggle_environments/envs/football/football.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.py:64]: Yes, I'd like to get those logs stored somewhere.
As we can use them to generate videos, debug issues etc.

AFAIK, this code will be run only on the 'orchestrator', we should not need to collect any info from the agent machines.

File: kaggle_environments/envs/football/test_football.py (hunk 1 above)
Comment (cyfra) [reply kaggle_environments/envs/football/test_football.py:30]: Yes, the environment has its own test suite.

Here, I wanted to only focus on testing the 'integration' part.
And yes, I plan to add more tests as we go.

File: kaggle_environments/__init__.py [hunk 12]
```diff
@@ -36,5 +36,6 @@
             "renderer": getattr(env, "renderer"),
             "specification": getattr(env, "specification"),
         })
-    except:
+    except Exception as e:
+        print("FAiled: %s: %s" % (name, e))
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/__init__.py:40]: done

File: kaggle_environments/envs/football/football.ipynb [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "## Installing kaggle-environments.\n",
+        "\n",
+        "!curl -X PURGE https://pypi.org/simple/kaggle-environments\n",
+        "\n",
+        "## Use the freshest version\n",
+        "!pip install git+https://github.com/cyfra/kaggle-environments@football_v1\n",
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.ipynb:28]: done

File: kaggle_environments/envs/football/football.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    for (x = 0; x<96; x++) {
+      for (z = 0; z < 4; z++) {
+        if (minimap[i] > 0) {
+          drawCell({x:x, y:y, color:colors[z]});
+        }
+        i += 1;
+      }
+    }
+
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.js]: done

File: kaggle_environments/envs/football/football.json [hunk 15]
```diff
@@ -0,0 +1,61 @@
+{
+    "name": "football",
+    "agents": [2],
+    "configuration": {
+      "episodeSteps": 30,
+      "scenario_name": {
+          "description": "Name of the scenario: for example 11_vs_11_stochastic. Look inside scenarios/ dir of GoogleResearchFootball.",
```
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.json]: done.

File: kaggle_environments/envs/football/football.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (cyfra) [reply kaggle_environments/envs/football/football.json]: done.

File: kaggle_environments/envs/football/football.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (harrisse) [reply kaggle_environments/envs/football/football.py:7]: Ah okay, I'm not sure I understand the model for how rendering will work, so we should chat about that tomorrow.
